File: Emerald.Streaming/ObsRecordingService.cs
using System.Diagnostics;

namespace Emerald.Streaming;

public sealed class ObsRecordingService : IDisposable
{
    private readonly object gate = new();
    private readonly string recordingsPath;
    private Process? process;

    public ObsRecordingService(string recordingsPath)
    {
        this.recordingsPath = recordingsPath;
    }

    public ObsRecordingStatus Status { get; private set; } = new(false, null, null, null, 120, "mp4", null);

    public ObsRecordingStatus Start(ObsRecordingRequest request)
    {
        if (string.IsNullOrWhiteSpace(request.InputUrl))
        {
            throw new InvalidOperationException("OBS recording URL is required.");
        }

        lock (gate)
        {
            if (IsProcessRunning())
            {
                return Status;
            }

            var segmentSeconds = Math.Clamp(request.SegmentSeconds ?? 120, 10, 3600);
            var container = NormalizeContainer(request.Container);
            var ffmpegPath = NormalizeFfmpegPath(request.FfmpegPath);
            var outputPattern = Path.Combine(recordingsPath, $"obs-%Y%m%d-%H%M%S.{container.Extension}");

            var startInfo = new ProcessStartInfo
            {
                FileName = ffmpegPath,
                UseShellExecute = false,
                RedirectStandardInput = true,
                RedirectStandardError = true,
                RedirectStandardOutput = true,
                CreateNoWindow = true
            };

            AddArguments(startInfo, request.InputUrl.Trim(), segmentSeconds, container.Format, outputPattern);

            var ffmpeg = new Process { StartInfo = startInfo, EnableRaisingEvents = true };
            ffmpeg.ErrorDataReceived += (_, args) =>
            {
                if (!string.IsNullOrWhiteSpace(args.Data))
                {
                    Status = Status with { LastMessage = args.Data };
                }
            };
            ffmpeg.Exited += (_, _) =>
            {
                Status = Status with { IsRecording = false, LastMessage = Status.LastMessage ?? "FFmpeg stopped." };
            };

            try
            {
                if (!ffmpeg.Start())
                {
                    throw new InvalidOperationException("Unable to start FFmpeg.");
                }
            }
            catch (Exception exception)
            {
                ffmpeg.Dispose();
                process = null;
                Status = Status with { IsRecording = false, LastMessage = exception.Message };
                throw new InvalidOperationException($"Unable to start FFmpeg at '{ffmpegPath}'. Use the full path to ffmpeg.exe or a folder that contains ffmpeg.exe.", exception);
            }

            process = ffmpeg;

            try
            {
                process.BeginErrorReadLine();
                process.BeginOutputReadLine();
            }
            catch (InvalidOperationException)
            {
                // Some process failures can exit before async stream readers attach.
            }

            Status = new ObsRecordingStatus(
                true,
                DateTimeOffset.UtcNow,
                request.InputUrl.Trim(),
                outputPattern,
                segmentSeconds,
                container.Extension,
                null);

            return Status;
        }
    }

    public ObsRecordingStatus Stop()
    {
        lock (gate)
        {
            if (process is null)
            {
                Status = Status with { IsRecording = false };
                return Status;
            }

            if (!process.HasExited)
            {
                try
                {
                    process.StandardInput.WriteLine("q");
                    if (!process.WaitForExit(5000))
                    {
                        process.Kill(entireProcessTree: true);
                    }
                }
                catch
                {
                    if (!process.HasExited)
                    {
                        process.Kill(entireProcessTree: true);
                    }
                }
            }

            process.Dispose();
            process = null;
            Status = Status with { IsRecording = false, LastMessage = "Recording stopped." };
            return Status;
        }
    }

    public void Dispose()
    {
        Stop();
    }

    private static void AddArguments(ProcessStartInfo startInfo, string inputUrl, int segmentSeconds, string segmentFormat, string outputPattern)
    {
        startInfo.ArgumentList.Add("-hide_banner");
        startInfo.ArgumentList.Add("-loglevel");
        startInfo.ArgumentList.Add("warning");
        startInfo.ArgumentList.Add("-i");
        startInfo.ArgumentList.Add(inputUrl);
        startInfo.ArgumentList.Add("-map");
        startInfo.ArgumentList.Add("0");
        startInfo.ArgumentList.Add("-c");
        startInfo.ArgumentList.Add("copy");
        startInfo.ArgumentList.Add("-f");
        startInfo.ArgumentList.Add("segment");
        startInfo.ArgumentList.Add("-segment_time");
        startInfo.ArgumentList.Add(segmentSeconds.ToString());
        startInfo.ArgumentList.Add("-reset_timestamps");
        startInfo.ArgumentList.Add("1");
        startInfo.ArgumentList.Add("-strftime");
        startInfo.ArgumentList.Add("1");
        startInfo.ArgumentList.Add("-segment_format");
        startInfo.ArgumentList.Add(segmentFormat);
        startInfo.ArgumentList.Add(outputPattern);
    }

    private static RecordingContainer NormalizeContainer(string? container)
    {
        return container?.Trim().ToLowerInvariant() switch
        {
            "mkv" or "matroska" => new RecordingContainer("mkv", "matroska"),
            "ts" or "mpegts" => new RecordingContainer("ts", "mpegts"),
            _ => new RecordingContainer("mp4", "mp4")
        };
    }

    private bool IsProcessRunning()
    {
        if (process is null)
        {
            return false;
        }

        try
        {
            if (!process.HasExited)
            {
                return true;
            }

            process.Dispose();
            process = null;
            return false;
        }
        catch (InvalidOperationException)
        {
            var failedProcess = process;
            failedProcess?.Dispose();
            process = null;
            Status = Status with { IsRecording = false };
            return false;
        }
    }

    private static string NormalizeFfmpegPath(string? configuredPath)
    {
        if (string.IsNullOrWhiteSpace(configuredPath))
        {
            return "ffmpeg";
        }

        var trimmedPath = configuredPath.Trim().Trim('"');

        if (Directory.Exists(trimmedPath))
        {
            return Path.Combine(trimmedPath, OperatingSystem.IsWindows() ? "ffmpeg.exe" : "ffmpeg");
        }

        return trimmedPath;
    }

    private sealed record RecordingContainer(string Extension, string Format);
}

public sealed record ObsRecordingRequest(
    string InputUrl,
    int? SegmentSeconds,
    string? Container,
    string? FfmpegPath);

public sealed record ObsRecordingStatus(
    bool IsRecording,
    DateTimeOffset? StartedAt,
    string? InputUrl,
    string? OutputPattern,
    int SegmentSeconds,
    string Container,
    string? LastMessage);

File: emerald-streaming-windows/publish/services/obsRecordingService.js
const fs = require("node:fs");
const os = require("node:os");
const path = require("node:path");
const { spawn } = require("node:child_process");

class ObsRecordingService {
  constructor(recordingsPath) {
    this.recordingsPath = recordingsPath;
    this.process = null;
    this.status = {
      isRecording: false,
      startedAt: null,
      inputUrl: null,
      outputPattern: null,
      segmentSeconds: 120,
      container: "mp4",
      lastMessage: null,
    };
  }

  start(request) {
    if (!request.inputUrl || !String(request.inputUrl).trim()) {
      throw new Error("Recording URL is required.");
    }

    if (this.isProcessRunning()) {
      return this.status;
    }

    const segmentSeconds = clamp(Number(request.segmentSeconds || 120), 10, 3600);
    const container = normalizeContainer(request.container);
    const ffmpegPath = normalizeFfmpegPath(request.ffmpegPath);
    const outputPattern = path.join(this.recordingsPath, `emerald-%Y%m%d-%H%M%S.${container.extension}`);
    const args = [
      "-hide_banner",
      "-loglevel", "warning",
      "-i", String(request.inputUrl).trim(),
      "-map", "0",
      "-c", "copy",
      "-f", "segment",
      "-segment_time", String(segmentSeconds),
      "-reset_timestamps", "1",
      "-strftime", "1",
      "-segment_format", container.format,
      outputPattern,
    ];

    try {
      this.process = spawn(ffmpegPath, args, {
        windowsHide: true,
        stdio: ["pipe", "pipe", "pipe"],
      });
    } catch (error) {
      this.process = null;
      this.status = { ...this.status, isRecording: false, lastMessage: error.message };
      throw new Error(`Unable to start FFmpeg at '${ffmpegPath}'. Use the full path to ffmpeg.exe or a folder that contains ffmpeg.exe.`);
    }

    this.process.stderr.on("data", (chunk) => {
      const message = chunk.toString().trim();
      if (message) {
        this.status = { ...this.status, lastMessage: message };
      }
    });

    this.process.on("error", (error) => {
      this.status = { ...this.status, isRecording: false, lastMessage: error.message };
    });

    this.process.on("exit", () => {
      this.status = { ...this.status, isRecording: false, lastMessage: explainFfmpegMessage(this.status.lastMessage || "FFmpeg stopped.") };
      this.process = null;
    });

    this.status = {
      isRecording: true,
      startedAt: new Date().toISOString(),
      inputUrl: String(request.inputUrl).trim(),
      outputPattern,
      segmentSeconds,
      container: container.extension,
      lastMessage: null,
    };

    return this.status;
  }

  stop() {
    if (!this.process) {
      this.status = { ...this.status, isRecording: false };
      return this.status;
    }

    const processToStop = this.process;

    if (!processToStop.killed) {
      try {
        processToStop.stdin.write("q\n");
      } catch {
        // Fall back to terminating below.
      }

      setTimeout(() => {
        if (this.process === processToStop && !processToStop.killed) {
          processToStop.kill("SIGKILL");
        }
      }, 5000).unref();
    }

    this.process = null;
    this.status = { ...this.status, isRecording: false, lastMessage: "Recording stopped." };
    return this.status;
  }

  isProcessRunning() {
    return Boolean(this.process && !this.process.killed);
  }
}

function normalizeContainer(container) {
  switch (String(container || "").trim().toLowerCase()) {
    case "mkv":
    case "matroska":
      return { extension: "mkv", format: "matroska" };
    case "ts":
    case "mpegts":
      return { extension: "ts", format: "mpegts" };
    default:
      return { extension: "mp4", format: "mp4" };
  }
}

function normalizeFfmpegPath(configuredPath) {
  if (!configuredPath || !String(configuredPath).trim()) {
    return "ffmpeg";
  }

  const trimmedPath = String(configuredPath).trim().replace(/^"|"$/g, "");

  if (fs.existsSync(trimmedPath) && fs.statSync(trimmedPath).isDirectory()) {
    return path.join(trimmedPath, os.platform() === "win32" ? "ffmpeg.exe" : "ffmpeg");
  }

  return trimmedPath;
}

function clamp(value, min, max) {
  if (Number.isNaN(value)) {
    return min;
  }

  return Math.min(Math.max(value, min), max);
}

module.exports = {
  ObsRecordingService,
  normalizeFfmpegPath,
};

function explainFfmpegMessage(message) {
  if (message && message.toLowerCase().includes("error opening input")) {
    return `${message} Check that the stream is streaming to rtmp://127.0.0.1:1935/live with stream key emerald, then start recording again.`;
  }

  return message;
}
